lib/spack/spack/modules/common.py

```diff
@@ -725,9 +726,34 @@ def autoload(self):
         return specs + literals
 
     def _create_module_list_of(self, what):
-        m = self.conf.module
-        return [m.make_layout(x).use_name
-                for x in getattr(self.conf, what)]
+        mod = self.conf.module
+        kind = mod.__name__.rsplit('.', 1)[-1]
+        index = dict()
+
+        def _load_indices(s):
+            if len(index):
+                return
+            root = mod.make_layout(s).dirname()
+            index.update(read_module_index(root))
+            for ups in read_module_indices():
+                index.update(ups.get(kind, {}))
+
+        def _valid(spec):
+            if not spec.external:
```

Comment:
I'd consider reducing the indents:
```
if spec.external:
    return True
_load_indices(....)
```

Reply:
Was thinking about that, at the additional cost of another `return True`. But is maybe more explicit that external modules will always be considered